Implement Targeted Search Options on Title List

carolegodfrey opened the pull request:
## Purpose
Implement Targeted Search Options for Title Searches.  This allows searching for titles by
- title (ie name)
- issn/isbn
- publisher
- subject
https://issues.folio.org/browse/UIEH-120

## Approach
- Search form is updated to use searchfield component ( (combines a select and text box) for Title searches. Continue to use simple search text box for Provider and Package searches
- eHoldings URL continues to use **q** parameter and adds a new **searchfield** parameter for Title search options, For example?q=education&searchfield=subject
- Supporting requests to mod-kb-ebsco are made using a JSON API filter syntax (for example, filter[subject]=education)
- Updates to mirage configuration to support new search filters
- When switching between search options (for example from publisher to subject), the search box maintains the search term

## Learning
- Used searchfield component as is used in codex search application
https://github.com/folio-org/stripes-components/blob/master/lib/structures/SearchField/readme.md
- https://www.npmjs.com/package/qs


## Screenshots
![targeted-search-revised](https://user-images.githubusercontent.com/19415226/36455517-73116818-166e-11e8-9f00-dc0063d6c72a.gif)

wwilsman commented on mirage/config.js:
```diff
@@ -9,18 +9,18 @@ import { Response } from 'mirage-server';
-function searchRouteFor(resourceType, filter = () => true) {
+function searchRouteFor(resourceType, filter = (title, req) => {
+  let query = req.queryParams.q.toLowerCase();
+  return title.name && title.name.toLowerCase().includes(query.toLowerCase());
```
To avoid confusion, we should use `model` here instead of `title` since this applies to packages and providers.

Also `let query` already has `toLowerCase` so the second `query.toLowerCase()` isn't necessary.

wwilsman commented on mirage/config.js:
```diff
@@ -226,9 +247,14 @@ export default function configure() {
+    return titles.find(request.params.id);
+  });
+
+
```
Looks like this was duplicated by accident

carolegodfrey replied:
Yes - probably an issue during a rebase

wwilsman commented on src/routes/search.js:
```diff
@@ -75,6 +74,21 @@ class SearchRoute extends Component {
 
+  /**
+   * Transforms UI params into search params
+   * @param {*} parms - query param object
```
This should be `@param {Object} params`

wwilsman commented on src/routes/search.js:
```diff
@@ -83,12 +97,13 @@ class SearchRoute extends Component {
     let { searchType, params } = this.state;
-    let { offset = 0, ...queryParams } = params;
+    let searchParams = this.transformQueryParams(params);
+    let { offset = 0, ...pagedParams } = searchParams;
```
`offset` comes from the URL params, so it shouldn't come from the tranformed search params. You could leave the old destructuring and just transform `queryParams`

```
let { offset = 0, ...queryParams } = params;
let searchParams = this.transformQueryParams(queryParams);
```

wwilsman commented on src/routes/search.js:
```diff
@@ -83,12 +97,13 @@ class SearchRoute extends Component {
     return resolver.query(searchType, {
       filter: undefined,
-      ...queryParams,
+      ...pagedParams,
```
This can be `searchParams` once we change that transform. `pagedParams` is sort of a misnomer since it doesn't actually include any page parameter.

wwilsman commented on tests/package-search-test.js:
```diff
@@ -148,7 +148,7 @@ describeApplication('PackageSearch', () => {
 
       it('displays an empty search', () => {
-        expect(PackageSearchPage.$searchField).to.have.value('');
+        expect(PackageSearchPage.$titleSearchField).to.have.value('');
```
Since titles are resources in our app `titleSearchField` feels a little confusing. Maybe we should go with `nameSearchField` since all of our resources have `name` attributes

carolegodfrey replied:
Only thing here is that we are referring to the title search element from the package search. It seems like name would apply to package and provider in addition to title. We could call them titleSearchField, packageSearchField and providerSearchField distinctly in all pages provider-search.js, title-search.js and package-search.js

wwilsman replied:
Ah, I see. I think what confused me is in the gif we're calling the title's name field `title` so I had assumed `$titleSearchField` was meant to be the name search field.

wwilsman commented on tests/pages/title-search.js:
```diff
@@ -27,9 +27,11 @@ export default {
   get $searchField() {
+    return $('[data-test-title-search-field]').find('input[name="search"]');
+  },
+  get $otherSearchTypeSearchField() {
```
In `pages/package-search` we left this as `$searchField` and called the other one `$titleSearchField`. We should keep these consistent and decide which way we want them to be.

carolegodfrey replied:
could add a packageSearchField and a providerSearchField (these would be the same element)

wwilsman commented on tests/title-search-test.js:
```diff
@@ -9,16 +9,42 @@ describeApplication('TitleSearch', () => {
+    titles = this.server.createList('title', 3, 'withPackages', 'withSubjects', 'withIdentifiers', {
       name: i => `Title${i + 1}`,
-      publicationType: i => (i % 2 ? 'book' : 'journal')
+      publicationType: i => (i % 2 ? 'book' : 'journal'),
```
We use this `i % 2` statement multiple times now, we should add a comment explaining that only odd indexed items are given these alternate attributes.

wwilsman commented on mirage/config.js:
```diff
@@ -214,8 +218,25 @@ export default function configure() {
+    let isxn = params['filter[isxn]'];
+    let subject = params['filter[subject]'];
+    let publisher = params['filter[publisher]'];
+    let queryFiltered = false;
```
We should default this to `true` and remove the last `else` below. This keeps it consistent with `filtered` and says that we should return true unless a more specific filter is provided.

carolegodfrey replied:
Thanks for all the comments @wwilsman - will refine as suggested. As an added note - reviewed functionality with Khalilah (when switching between SearchOptions) and we will be maintaining the search term rather than clearing it out

carolegodfrey commented:
- Revised code to address requested code changes
- Corrected bug where an incorrect search filter (filter[name],filter[isxn],filter[publisher] or filter[subject])  was being propagated on url when switching search types (if both a search field and another filter were selected on a title search)
- Added unit tests to check bug fix 
- Re-Recorded gif to reflect requested  functionality - search box does no longer clear when changing search options
